File: 0379-Design Phone Directory/0379-Design Phone Directory.cpp
class PhoneDirectory {
public:
    /** Initialize your data structure here
        @param maxNumbers - The maximum numbers that can be stored in the phone directory. */
    PhoneDirectory(int maxNumbers) {
        used.resize(maxNumbers);
        dict.resize(maxNumbers);
        index = 0;
        for (int i = 0; i < maxNumbers; ++i)
        {
            dict[i] = i;
        }
    }
    
    /** Provide a number which is not assigned to anyone.
        @return - Return an available number. Return -1 if none is available. */
    int get() {
        if (index == used.size())
        {
            return -1;
        }
        
        int result = dict[index++];
        used[result] = true;
        return result;
    }
    
    /** Check if a number is available or not. */
    bool check(int number) {        
        return number >= 0 && number < used.size() && !used[number];
    }
    
    /** Recycle or release a number. */
    void release(int number) {
        if (check(number))
        {
            return;
        }
        
        used[number] = false;
        dict[--index] = number;
    }
    
private:
    vector<bool> used;
    vector<int> dict;
    int index;
};

/**
 * Your PhoneDirectory object will be instantiated and called as such:
 * PhoneDirectory obj = new PhoneDirectory(maxNumbers);
 * int param_1 = obj.get();
 * bool param_2 = obj.check(number);
 * obj.release(number);
 */
 

File: 0301-0400/0379-Design Phone Directory/0379-Design Phone Directory.py
class PhoneDirectory:
    
    def __init__(self, maxNumbers: int):
        """
        Initialize your data structure here
        @param maxNumbers - The maximum numbers that can be stored in the phone directory.
        """
        self.used = [False] * maxNumbers
        self.table = [i for i in range(maxNumbers)]
        self.index = 0
        

    def get(self) -> int:
        """
        Provide a number which is not assigned to anyone.
        @return - Return an available number. Return -1 if none is available.
        """
        if self.index == len(self.used):
            return -1
        result = self.table[self.index]
        self.index += 1
        self.used[result] = True
        return result
        

    def check(self, number: int) -> bool:
        """
        Check if a number is available or not.
        """
        return 0 <= number < len(self.used) and not self.used[number]
        

    def release(self, number: int) -> None:
        """
        Recycle or release a number.
        """
        if self.check(number):
            return
        self.used[number] = False
        self.index -= 1
        self.table[self.index] = number        


# Your PhoneDirectory object will be instantiated and called as such:
# obj = PhoneDirectory(maxNumbers)
# param_1 = obj.get()
# param_2 = obj.check(number)
# obj.release(number)
